src/common/chunked.ts

```diff
@@ -0,0 +1,457 @@
+/**
+ * Copyright (C) 2022-2024 Permanent Data Solutions, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { CanceledError } from 'axios';
+import { pLimit } from 'plimit-lit';
+import { Readable } from 'stream';
+
+import {
+  ByteCount,
+  TurboChunkingMode,
+  TurboLogger,
+  TurboUploadDataItemResponse,
+  UploadSignedDataItemParams,
+  validChunkingModes,
+} from '../types.js';
+import { TurboEventEmitter, createStreamWithUploadEvents } from './events.js';
+import { TurboHTTPService } from './http.js';
+import { TurboWinstonLogger } from './logger.js';
+
+const fiveMiB = 5 * 1024 * 1024; // 5 MiB
+const fiveHundredMiB = fiveMiB * 100; // 500 MiB
+export const defaultMaxChunkConcurrency = 5; // Default max chunk concurrency
+
+export const maxChunkByteCount = fiveHundredMiB;
+export const minChunkByteCount = fiveMiB;
+export const defaultChunkByteCount = minChunkByteCount; // Default chunk size for uploads
+
+const backlogQueueFactor = 2;
+const chunkingHeader = { 'x-chunking-version': '2' } as const;
+
+/**
+ * Performs a chunked upload by splitting the stream into fixed-size buffers,
+ * uploading them in parallel, and emitting progress/error events.
+ */
+export class ChunkedUploader {
+  private chunkByteCount: number;
+  private readonly maxChunkConcurrency: number;
+  private readonly http: TurboHTTPService;
+  private readonly token: string;
+  private readonly logger: TurboLogger;
+
+  constructor({
+    http,
+    token,
+    maxChunkConcurrency = defaultMaxChunkConcurrency,
+    chunkByteCount = defaultChunkByteCount,
+    logger = TurboWinstonLogger.default,
+  }: {
+    http: TurboHTTPService;
+    token: string;
+    logger: TurboLogger;
+    chunkByteCount?: number;
```

Comment:
nit: can take chunkingMode here... and then on constructor assert params and set shouldChunk, expose "shouldChunk" as a public boolean for business logic. handles this more cleanly without statics